rlclientlib/logger/logger_facade.cc

```diff
@@ -92,19 +77,39 @@ namespace reinforcement_learning {
       perror_cb) : nullptr) {
     }
 
-    int slates_logger_facade::init(api_status* status) {
+    int multi_slot_logger_facade::init(api_status* status) {
       switch (_version) {
         case 1: return _v1->init(status);
         case 2: return _v2->init(status);
         default: return protocol_not_supported(status);
       }
     }
 
-    int slates_logger_facade::log_decision(const std::string& event_id, const char* context, unsigned int flags, const std::vector<std::vector<uint32_t>>& action_ids,
+    int multi_slot_model_type_to_payload_type(model_type model_type, generic_event::payload_type_t& payload_type, api_status* status)
+    {
+      //XXX out params must be always initialized. This is an ok default
+      payload_type = generic_event::payload_type_t::PayloadType_Slates;
+      switch(model_type) {
+        case reinforcement_learning::model_management::model_type_t::CCB: payload_type = generic_event::payload_type_t::PayloadType_CCB; break;
+        case reinforcement_learning::model_management::model_type_t::SLATES: payload_type = generic_event::payload_type_t::PayloadType_Slates; break;
+        default: RETURN_ERROR_ARG(nullptr, status, invalid_argument, "Invalid model_type, only Slates and CCB are supported with multi_slot decisions");
+      }
+      return error_code::success;
+    }
+
+    int multi_slot_logger_facade::log_decision(model_type model_type, const std::string& event_id, const char* context, unsigned int flags, const std::vector<std::vector<uint32_t>>& action_ids,
       const std::vector<std::vector<float>>& pdfs, const std::string& model_version, api_status* status) {
       switch (_version) {
-        case 1: return _v1->log_decision(event_id, context, flags, action_ids, pdfs, model_version, status);
-        case 2: return _v2->log(event_id.c_str(), _serializer.event(context, flags, action_ids, pdfs, model_version), _serializer.type, status);
+        case 1: {
```

Comment:
same here, can we pass the model type during the logger object `init()` and make this check there once? from what I understand it's not like we can continue execution if we are using the wrong protocol version with the wrong logger, right?

Reply:
agree, it seems better to make model_type as part of logger_facade state

Reply:
We can't check during initialization because it's not conditional. multi_slot_logger_facade is always initialized, regardless of the underlying model.

We can move the model type probing to construction, but that doesn't really buy us much.

Reply:
That's actually not great. Every logger initialization contributes to amount of background threads. Now we have model_type explosed, so we can fix it: combination of model_type + version clearly identify the only one valid rank call, so we can instantiate only one required underlying logger during init call (and throw error for incompatible combination).
It seems like other step of refactoring though, so we can postpone it to next PR (happy to do it).